taichi/codegen/spirv/spirv_ir_builder.cpp

```diff
@@ -1168,17 +1169,31 @@ void IRBuilder::init_random_function(Value global_tmp_) {
   store_var(rand_y_, _362436069u);
   store_var(rand_z_, _521288629u);
   store_var(rand_w_, _88675123u);
-  // Yes, this is not an atomic operation, but just fine since no matter
-  // how RAND_STATE changes, `gl_GlobalInvocationID.x` can still help
-  // us to set different seeds for different threads.
-  // Discussion:
-  // https://github.com/taichi-dev/taichi/pull/912#discussion_r419021918
-  Value tmp9 = load_var(rand_gtmp_, t_uint32_);
-  Value tmp10 = new_value(t_uint32_, ValueKind::kNormal);
-  ib_.begin(spv::OpIAdd)
-      .add_seq(t_uint32_, tmp10, tmp9, _1)
-      .commit(&func_header_);
-  store_var(rand_gtmp_, tmp10);
+
+  enum spv::Op add_op = spv::OpIAdd;
+
+  // use atomic iadd for DX API otherwise the HLSL compiler throws error:
+  // error X3694: race condition writing to shared resource detected
+  if (dynamic_cast<const taichi::lang::directx11::Dx11Device *>(device_)) {
```

Comment:
Some how I remember fixing this problem before but don't know why it is reverted later. My bad. This should work for now!

Reply:
Should guard this if statement under `#ifdef TI_WITH_DX11` as well

Reply:
Done, no problem it gave me a good opportunity to learn a bit more about the codegen :D